sdlib/dmd/gc.d

```diff
@@ -67,48 +67,45 @@ bool __sd_gc_fetch_alloc_info(void* ptr, void** base, size_t* size,
 	return true;
 }
 
-size_t __sd_gc_get_array_used(void* ptr, PageDescriptor pd) {
-	auto e = pd.extent;
-	if (!e) {
-		return 0;
-	}
-
-	if (pd.isSlab()) {
-		auto si = SlabAllocInfo(pd, ptr);
-		return si.usedCapacity;
-	} else {
-		return e.usedCapacity;
-	}
-}
-
-// TODO: this should only do large blocks, and let the druntime side do slabs.
-bool __sd_gc_set_array_used(void* ptr, PageDescriptor pd, size_t newUsed,
-                            size_t existingUsed) {
+bool __sd_gc_shrink_array_used(void* ptr, size_t newUsed, size_t existingUsed) {
+	auto pd = threadCache.maybeGetPageDescriptor(ptr);
 	auto e = pd.extent;
 	if (!e) {
 		return false;
 	}
 
 	if (pd.isSlab()) {
 		auto si = SlabAllocInfo(pd, ptr);
-		if (existingUsed < size_t.max && existingUsed != si.usedCapacity) {
-			// The existing capacity doesn't match.
+		auto offset = ptr - si.address + 1;
+		if (existingUsed != 0 && !threadCache
```

Comment:
Why do you check for `existingUsed != 0` here?

Reply:
For assumeSafeAppend. I don't know the existing used space.

Reply:
I suppose I could always call the getslice function.

Reply:
That is what it is for :)

Reply:
It's only needed if the code needs to run finalizers on "removed" elements, but now it will be needed for every case. We can see how it might affect performance.